packages/@aws-cdk/aws-codepipeline/lib/pipeline.ts

```diff
@@ -445,6 +415,93 @@ export class Pipeline extends PipelineBase {
     return actionRole;
   }
 
+  private getRoleFromActionPropsOrGenerateIfCrossAccount(stage: Stage, action: IAction): iam.IRole | undefined {
+    if (action.actionProperties.role) {
+      if (this.isAwsOwned(action)) {
+        return action.actionProperties.role;
+      } else {
+        throw new Error("Specifying a Role is not supported for actions with an owner different than 'AWS' - " +
+          `got '${action.actionProperties.owner}' (Action: '${action.actionProperties.actionName}' in Stage: '${stage.stageName}')`);
+      }
+    }
+
+    const otherAccountStack = this.getOtherStackIfActionIsCrossAccount(action);
+    if (!otherAccountStack) {
+      return undefined;
+    }
+
+    // if we have a cross-account action, the pipeline's bucket must have a KMS key
+    if (!this.artifactBucket.encryptionKey) {
+      throw new Error('The Pipeline is being used in a cross-account manner, ' +
+        'but its artifact bucket does not have a KMS key defined. ' +
+        'A KMS key is required for a cross-account Pipeline. ' +
+        'Make sure to pass a Bucket with a Key when creating the Pipeline');
+    }
+
+    // generate a role in the other stack, that the Pipeline will assume for executing this action
+    const pipelineStack = Stack.of(this);
+    const ret = new iam.Role(otherAccountStack,
+        `${this.node.uniqueId}-${stage.stageName}-${action.actionProperties.actionName}-ActionRole`, {
+      assumedBy: new iam.AccountPrincipal(pipelineStack.account),
+      roleName: PhysicalName.GENERATE_IF_NEEDED,
+    });
+    // the other stack with the role has to be deployed before the pipeline stack
+    pipelineStack.addDependency(otherAccountStack);
+
+    return ret;
+  }
+
+  /**
+   * Returns the Stack this Action belongs to if this is a cross-account Action.
+   * If this Action is not cross-account (i.e., it lives in the same account as the Pipeline),
+   * it returns undefined.
+   *
+   * @param action the Action to return the Stack for
+   */
+  private getOtherStackIfActionIsCrossAccount(action: IAction): Stack | undefined {
+    const pipelineStack = Stack.of(this);
+
+    if (action.actionProperties.resource) {
+      const resourceStack = Stack.of(action.actionProperties.resource);
+      // check if resource is from a different account
+      return pipelineStack.account === resourceStack.account
+        ? undefined
+        : resourceStack;
+    }
+
+    if (!action.actionProperties.account) {
+      return undefined;
+    }
+
+    const targetAccount = action.actionProperties.account;
+    // check whether the account is a static string
+    if (Token.isUnresolved(targetAccount)) {
+      throw new Error(`Cannot use dynamic value for account! (action: '${action.actionProperties.actionName}')`);
```

Comment:
no "!" in messages

Reply:
Also the word "dynamic" is not the terminology we normally use. Maybe something like "stack specify a concrete account in "env""

Reply:
Fixed.